src/perf_counters.cc

```diff
@@ -26,9 +26,121 @@
 namespace benchmark {
 namespace internal {
 
-constexpr size_t PerfCounterValues::kMaxCounters;
-
 #if defined HAVE_LIBPFM
+
+class SinglePMURegistry {
+ public:
+  ~SinglePMURegistry() = default;
+  SinglePMURegistry(SinglePMURegistry&&) = default;
+  SinglePMURegistry(const SinglePMURegistry&) = delete;
+  SinglePMURegistry& operator=(SinglePMURegistry&&) noexcept;
+  SinglePMURegistry& operator=(const SinglePMURegistry&) = delete;
+
+  SinglePMURegistry(pfm_pmu_t pmu_id)
+      : pmu_id_(pmu_id), available_counters_(0), available_fixed_counters_(0) {
+    {
+      pfm_pmu_info_t pmu_info{};
+      const auto pfm_pmu = pfm_get_pmu_info(pmu_id, &pmu_info);
+
+      if (pfm_pmu != PFM_SUCCESS) {
+        GetErrorLogInstance() << "Unknown pmu: " << pmu_id << "\n";
+        return;
+      }
+
+      name_ = pmu_info.name;
+      desc_ = pmu_info.desc;
+      available_counters_ = pmu_info.num_cntrs;
+      available_fixed_counters_ = pmu_info.num_fixed_cntrs;
+
+      BM_VLOG(1) << "PMU: " << pmu_id << " " << name_ << " " << desc_ << "\n";
+      BM_VLOG(1) << "     counters: " << available_counters_ << " fixed: " << available_fixed_counters_ << "\n";
+    }
+  }
+
+  const char* name() const { return name_; }
+
+  bool AddCounter(int event_id) {
+    pfm_event_info_t info{};
+    const auto pfm_event_info =
+        pfm_get_event_info(event_id, PFM_OS_PERF_EVENT, &info);
+
+    if (pfm_event_info != PFM_SUCCESS) {
+      GetErrorLogInstance() << "Unknown event id: " << event_id << "\n";
+      return false;
+    }
+
+    assert(info.pmu == pmu_id_);
+
+    if (counter_ids_.find(event_id) != counter_ids_.end()) return true;
+
+    if (counter_ids_.size() >= available_counters_ - 1) {
+      GetErrorLogInstance() << "Maximal number of counters for PMU " << name_
+                            << " (" << available_counters_ << ") reached.\n";
+      return false;
+    }
+
+    counter_ids_.emplace(event_id, info.code);
+
+    BM_VLOG(2) << "Registered counter: " << event_id << " (" << info.name << " - " << info.desc
+               << ") in pmu " << name_ << " (" << counter_ids_.size() << "/" << available_counters_ << "\n";
+
+    return true;
+  }
+
+ private:
+  pfm_pmu_t pmu_id_;
+  const char* name_;
+  const char* desc_;
+  std::unordered_map<int, uint64_t> counter_ids_;
+  std::unordered_map<int, uint64_t> fixed_counter_ids_;
+  uint64_t available_counters_;
```

Comment:
https://google.github.io/styleguide/cppguide.html#Integer_Types

for pretty good reason, we try to stick to `int` rather than `uint64_t`.  obviously if 64-bit is required that is fine, but please don't use unsigned for these.  instead used signed and check for over/underflow instead.

Reply:
Well, in this case, I'm just trying to stick to whatever libpfm does:

```c
typedef struct {
	...
	uint64_t		code;	/* event raw code (not encoding) */
    ...
} pfm_event_info_t;
```

Do you prefer something else given their choice of types?

Reply:
i think the map should stay as `uint64_t` as we have no way to know what they're using for ids (they could be 64-bit UUID or something).  the _counters_ should change to int though.

Reply:
So, this introduces a warning about comparting `.size()` which returns a `size_t` to `int` and the possible loss of integer precision. Do you want me to cast the `available_counters_` to `size_t` for the comparison, or just leave as is?

Reply:
ideally, we'd check the returned `.size()` is within bounds for int(64_t) and then cast to int(64_t).

i'm much more concerned about the unsigned bit than the 64-bit bit (though if we have more counters than fit in 32-bit... :O

Reply:
OK, This feels over the top to me considering the amount of `.size()` and the use of `size_t` everywhere, due to STL, but there you go, updated the code.